core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -139,11 +142,11 @@ class Partition(val topic: String,
   }
 
   def getOrCreateReplica(replicaId: Int = localBrokerId, isNew: Boolean = false): Replica = {
-    assignedReplicaMap.getAndMaybePut(replicaId, {
+    allReplicasMap.getAndMaybePut(replicaId, {
```

Comment:
I think we actually read HW from the checkpoint if the replica is a future one. When the future replica is newly created, it is OK that we don't have HW for the future replica because the log for this future replica is empty. Then `ReplicaManager.checkpointHighWatermarks()` will checkpoint HW for the future replica in the destination log directory. When broker restarts and load the log, Partition.getOrCreateReplica() will read the HW checkpoint in the destination log directory for future replica as well.

Does this make sense?

Reply:
Thanks. Make sense. Missed that isReplicaLocal() covers future replica too.